crates/router/src/core/relay.rs

```diff
@@ -0,0 +1,221 @@
+use api_models::relay as relay_models;
+use common_utils::{
+    self,
+    ext_traits::OptionExt,
+    id_type::{self, GenerateId},
+};
+use error_stack::ResultExt;
+
+use super::errors::{self, ConnectorErrorExt, RouterResponse, RouterResult, StorageErrorExt};
+use crate::{
+    core::payments,
+    routes::SessionState,
+    services,
+    types::{
+        api::{self},
+        domain,
+        transformers::ForeignFrom,
+    },
+};
+
+pub mod utils;
+
+pub async fn relay(
+    state: SessionState,
+    merchant_account: domain::MerchantAccount,
+    profile_id_optional: Option<id_type::ProfileId>,
+    key_store: domain::MerchantKeyStore,
+    req: relay_models::RelayRequest,
+) -> RouterResponse<relay_models::RelayResponse> {
+    let db = state.store.as_ref();
+    let key_manager_state = &(&state).into();
+    let merchant_id = merchant_account.get_id();
+
+    let profile_id_from_auth_layer = profile_id_optional
+        .get_required_value("ProfileId")
+        .change_context(errors::ApiErrorResponse::MissingRequiredField {
+            field_name: "profile id",
+        })?;
+
+    let profile = db
+        .find_business_profile_by_merchant_id_profile_id(
+            key_manager_state,
+            &key_store,
+            merchant_id,
+            &profile_id_from_auth_layer,
+        )
+        .await
+        .change_context(errors::ApiErrorResponse::ProfileNotFound {
+            id: profile_id_from_auth_layer.get_string_repr().to_owned(),
+        })?;
+
+    let relay_response = match req.relay_type {
+        common_enums::RelayType::Refund => {
+            Box::pin(relay_refund(
+                state,
+                merchant_account,
+                profile,
+                key_store,
+                &req,
+            ))
+            .await?
+        }
+    };
+    Ok(hyperswitch_domain_models::api::ApplicationResponse::Json(
+        relay_response,
+    ))
+}
+
+pub async fn relay_refund(
+    state: SessionState,
+    merchant_account: domain::MerchantAccount,
+    profile: domain::Profile,
+    key_store: domain::MerchantKeyStore,
+    req: &relay_models::RelayRequest,
+) -> RouterResult<relay_models::RelayResponse> {
+    validate_relay_refund_request(req).attach_printable("Invalid relay refund request")?;
+
+    let db = state.store.as_ref();
+    let key_manager_state = &(&state).into();
+    let connector_id = &req.connector_id;
+
+    let merchant_id = merchant_account.get_id();
+
+    let connector_account = db
```

Comment:
move this function to the parent, fetching the connector_account is irrespective of relay type.